File: leetcode/rust/src/20.valid-parentheses.rs
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack = vec![];
        
        for ch in s.chars().into_iter(){
            match stack.last(){
                None=>{},
                Some(&last) =>{
                    if Solution::pair(last, ch){
                        stack.pop();
                        continue;
                    }
                    
                }
            }
            stack.push(ch);
        }
        stack.is_empty()
    }

     #[inline(always)]
    fn pair(lhs :char , rhs :char )->bool{
         (lhs == '{' && rhs == '}') ||
            (lhs == '(' && rhs == ')') ||
            (lhs == '[' && rhs == ']')
    }
}

File: leetcode/python/20.valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) == 0:
            return True
        if len(s) % 2 == 1:
            return False

        stack = []
        map = dict()
        map[')'] = '('
        map[']'] = '['
        map['}'] = '{'

        for i in s:
            if i not in map:
                stack.append(i)
            else:
                if len(stack) > 0 and stack[-1] == map[i]:
                    stack.pop()
                else:
                    return False

        return len(stack) == 0
